File: cli_terminal/input/templates.py
"""Command template management with parameter expansion."""

import re
from typing import Dict, List, Optional
from dataclasses import dataclass


@dataclass
class Template:
    """A command template with parameters."""
    command: str
    label: Optional[str] = None
    param_prompts: Optional[Dict[str, str]] = None

    @property
    def params(self) -> List[str]:
        """Extract parameter names from command."""
        matches = re.findall(r"\{(\w+)\}", self.command)
        seen = set()
        result = []
        for p in matches:
            if p not in seen:
                seen.add(p)
                result.append(p)
        return result

    def expand(self, values: Dict[str, str]) -> str:
        """Expand template with given values."""
        result = self.command
        for param, value in values.items():
            result = result.replace("{" + param + "}", value)
        return result


class TemplateManager:
    """Manages command templates."""

    def __init__(self, templates_config: List[Dict]):
        self._templates: List[Template] = []
        for cfg in templates_config:
            label = cfg.get("label", "")
            command = cfg.get("command", "")
            prompts = {}
            for param in cfg.get("params", []):
                prompts[param["name"]] = param.get("prompt", param["name"])
            self._templates.append(Template(
                command=command,
                label=label,
                param_prompts=prompts if prompts else None,
            ))

    def get(self, label: str) -> Optional[Template]:
        """Get template by label."""
        for t in self._templates:
            if t.label == label:
                return t
        return None

    @property
    def all(self) -> List[Template]:
        """Get all templates."""
        return self._templates.copy()

    def expand(self, template: Template, values: Dict[str, str]) -> str:
        """Expand a template with values."""
        return template.expand(values)

File: src/input/templates.rs
use regex::Regex;

use crate::config::manager::{TemplateConfig, TemplateParam};

/// A parsed command template.
pub struct Template {
    pub label: String,
    /// Raw command string with `{param}` placeholders.
    command: String,
    /// Ordered list of unique parameter names.
    pub params: Vec<TemplateParam>,
}

impl Template {
    /// Get the raw command string.
    pub fn command(&self) -> &str {
        &self.command
    }

    pub fn from_config(cfg: TemplateConfig) -> Self {
        // Deduplicate params by name, preserving order.
        let mut seen = std::collections::HashSet::new();
        let params: Vec<TemplateParam> = cfg
            .params
            .into_iter()
            .filter(|p| seen.insert(p.name.clone()))
            .collect();
        Self {
            label: cfg.label,
            command: cfg.command,
            params,
        }
    }

    /// Expand template with parameter values. Unmatched placeholders are left as-is.
    pub fn expand(&self, values: &[(String, String)]) -> String {
        let mut result = self.command.clone();
        for (name, value) in values {
            result = result.replace(&format!("{{{name}}}"), value);
        }
        result
    }

    /// Extract parameter names from the command string.
    pub fn extract_params(command: &str) -> Vec<String> {
        let re = Regex::new(r"\{(\w+)\}").unwrap();
        let mut seen = std::collections::HashSet::new();
        re.captures_iter(command)
            .filter_map(|cap| cap.get(1))
            .filter(|m| seen.insert(m.as_str().to_string()))
            .map(|m| m.as_str().to_string())
            .collect()
    }
}

/// Manager for command templates.
pub struct TemplateManager {
    templates: Vec<Template>,
}

impl TemplateManager {
    pub fn new(configs: Vec<TemplateConfig>) -> Self {
        Self {
            templates: configs.into_iter().map(Template::from_config).collect(),
        }
    }

    pub fn templates(&self) -> &[Template] {
        &self.templates
    }

    pub fn by_label(&self, label: &str) -> Option<&Template> {
        self.templates.iter().find(|t| t.label == label)
    }

    pub fn update(&mut self, configs: Vec<TemplateConfig>) {
        self.templates = configs.into_iter().map(Template::from_config).collect();
    }
}
